quickwit/quickwit-control-plane/src/debouncer.rs

```diff
@@ -0,0 +1,198 @@
+// Copyright (C) 2024 Quickwit, Inc.
+//
+// Quickwit is offered under the AGPL v3.0 and as commercial software.
+// For commercial licensing, contact us at hello@quickwit.io.
+//
+// AGPL:
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Affero General Public License as
+// published by the Free Software Foundation, either version 3 of the
+// License, or (at your option) any later version.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU Affero General Public License for more details.
+//
+// You should have received a copy of the GNU Affero General Public License
+// along with this program. If not, see <http://www.gnu.org/licenses/>.
+
+use std::sync::atomic::{AtomicU8, Ordering};
+use std::sync::Arc;
+use std::time::Duration;
+
+use quickwit_actors::{Actor, ActorContext, DeferableReplyHandler, Handler};
+
+/// A debouncer is a helper to debounce events.
+///
+/// The debouncing takes a `cool_down_period` parameters and works like you may expect:
+///
+///
+///    time                t=0                   t=COOLDOWN            t=2*COOLDOWN
+///                        |                          |                     |
+///  ----------------------------------------------------------------------------------------------------
+///    event               *     * *    *                                   *   * *
+///    debounced effect    o                          o                     o
+///
+/// In particular, note the event triggered at `t=COOLDOWN`.
+#[derive(Clone)]
+pub struct Debouncer {
+    cool_down_period: Duration,
+    cooldown_state: Arc<AtomicU8>,
+}
+
+impl Debouncer {
+    pub fn new(cool_down_period: Duration) -> Debouncer {
+        Debouncer {
+            cool_down_period,
+            cooldown_state: Arc::new(AtomicU8::new(NO_COOLDOWN)),
```

Comment:
Why not use an enum and a sync mutex?

Reply:
changed the implem to enum and sync mutex.

I also took it as an opportunity to change it into a proper state machine, with atomic transitions.